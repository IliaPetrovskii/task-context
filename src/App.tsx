import { useState } from 'react';
import { data, IItem } from './data';
import './styles.css';

type Theme = 'light' | 'dark';

export function App() {
    const [currentTheme, setCurrentTheme] = useState<Theme>('light');

    function changeTheme() {
        setCurrentTheme(currentTheme === 'light' ? 'dark' : 'light');
    }

    const className = `app app_${currentTheme}`;
    return (
        <div className={className}>
            <button onClick={changeTheme}>Toggle theme</button>
            <List theme={currentTheme} data={data} />
        </div>
    );
}

function List(props: { theme: Theme; data: IItem[] }) {
    return (
        <div>
            {data.map((item) => (
                <ListItem
                    theme={props.theme}
                    caption={item.name}
                    key={item.id}
                />
            ))}
        </div>
    );
}

function ListItem(props: { theme: Theme; caption: string }) {
    const className = `listItem listItem_${props.theme}`;
    return <div className={className}>{props.caption}</div>;
}
