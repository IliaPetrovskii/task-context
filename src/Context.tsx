import { ReactElement } from 'react';

type Theme = 'light' | 'dark';

export function ThemeProvider(props: {
    theme: Theme;
    children: ReactElement;
}) {}

export function useTheme(): Theme {}
