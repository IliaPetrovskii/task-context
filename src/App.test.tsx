import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { App } from './App';
import { ThemeProvider, useTheme } from './Context';

describe('Контекст', () => {
    it('Список правильно отрисовывается и реагирует на переключение темы', () => {
        const { container } = render(<App />);

        expect(container).toMatchSnapshot();

        const btn = screen.getByText('Toggle theme');
        userEvent.click(btn);

        expect(container).toMatchSnapshot();
    });

    it('ThemeProvider спускает тему через контекст, useTheme получает тему из контекста', () => {
        function Consumer() {
            const theme = useTheme();
            return <div>{theme}</div>;
        }

        const { container, rerender } = render(
            <ThemeProvider theme={'dark'}>
                <Consumer />
            </ThemeProvider>,
        );

        expect(container).toMatchSnapshot();

        rerender(
            <ThemeProvider theme={'light'}>
                <Consumer />
            </ThemeProvider>,
        );

        expect(container).toMatchSnapshot();
    });
});
