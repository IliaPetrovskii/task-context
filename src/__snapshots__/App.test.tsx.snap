// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Контекст ThemeProvider спускает тему через контекст, useTheme получает тему из контекста 1`] = `
<div>
  <div>
    dark
  </div>
</div>
`;

exports[`Контекст ThemeProvider спускает тему через контекст, useTheme получает тему из контекста 2`] = `
<div>
  <div>
    light
  </div>
</div>
`;

exports[`Контекст Список правильно отрисовывается и реагирует на переключение темы 1`] = `
<div>
  <div
    class="app app_light"
  >
    <button>
      Toggle theme
    </button>
    <div>
      <div
        class="listItem listItem_light"
      >
        Иванов И.И.
      </div>
      <div
        class="listItem listItem_light"
      >
        Петров П.П.
      </div>
      <div
        class="listItem listItem_light"
      >
        Сидоров С.С.
      </div>
    </div>
  </div>
</div>
`;

exports[`Контекст Список правильно отрисовывается и реагирует на переключение темы 2`] = `
<div>
  <div
    class="app app_dark"
  >
    <button>
      Toggle theme
    </button>
    <div>
      <div
        class="listItem listItem_dark"
      >
        Иванов И.И.
      </div>
      <div
        class="listItem listItem_dark"
      >
        Петров П.П.
      </div>
      <div
        class="listItem listItem_dark"
      >
        Сидоров С.С.
      </div>
    </div>
  </div>
</div>
`;
